internal/notifier/github_helpers.go

```diff
@@ -0,0 +1,156 @@
+	if len(token) == 0 {
+		githubOpts, err := getGitHubAppOptions(providerName, providerNamespace, proxyURL, secretData, tokenCache)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(githubOpts) == 0 {
+			return nil, nil, errors.New("github token or github app details must be specified")
+		}
+
+		client, err := authgithub.New(githubOpts...)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		appToken, err := client.GetToken(context.Background())
+		if err != nil {
+			return nil, nil, err
+		}
+		token = appToken.Token
+	}
+
+	host, id, err := parseGitAddress(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	comp := strings.Split(id, "/")
+	if len(comp) != 2 {
+		return nil, nil, fmt.Errorf("invalid repository id %q", id)
+	}
+
+	baseUrl, err := url.Parse(host)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
+	tc := oauth2.NewClient(context.Background(), ts)
+	client := gogithub.NewClient(tc)
+	if baseUrl.Host != "github.com" {
+		if certPool != nil {
+			tr := &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs: certPool,
+				},
+			}
+			hc := &http.Client{Transport: tr}
+			ctx := context.WithValue(context.Background(), oauth2.HTTPClient, hc)
+			tc = oauth2.NewClient(ctx, ts)
+		}
+		client, err = gogithub.NewClient(tc).WithEnterpriseURLs(host, host)
+		if err != nil {
+			return nil, nil, fmt.Errorf("could not create enterprise GitHub client: %v", err)
+		}
+	}
+
+	return comp, client, nil
+}
+
+// getGithubClient constructs and returns the Github object.
+func getGitHubClient(commitStatus string, addr string, token string, certPool *x509.CertPool, proxyURL string, providerName string, providerNamespace string, secretData map[string][]byte, tokenCache *pkgcache.TokenCache) (*GitHub, error) {
+	repoInfo, client, err := getRepoInfoAndGithubClient(addr, token, certPool, proxyURL, providerName, providerNamespace, secretData, tokenCache)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GitHub{
+		Owner:        repoInfo[0],
+		Repo:         repoInfo[1],
+		CommitStatus: commitStatus,
+		Client:       client,
+	}, nil
+}
+
+// getGitHubDispatchClient constructs and returns the GithubDispatch object.
+func getGitHubDispatchClient(addr string, token string, certPool *x509.CertPool, proxyURL string, providerName string, providerNamespace string, secretData map[string][]byte, tokenCache *pkgcache.TokenCache) (*GitHubDispatch, error) {
+	repoInfo, client, err := getRepoInfoAndGithubClient(addr, token, certPool, proxyURL, providerName, providerNamespace, secretData, tokenCache)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GitHubDispatch{
+		Owner:  repoInfo[0],
+		Repo:   repoInfo[1],
+		Client: client,
+	}, nil
+}
```

Comment:
These two functions can be inlined in the two constructors since they are specific to each provider.